src/coreclr/src/vm/siginfo.cpp

```diff
@@ -3567,6 +3567,346 @@ BOOL CompareTypeTokens(mdToken tk1, mdToken tk2, Module *pModule1, Module *pModu
+    PCCOR_SIGNATURE     pEndSig,
+    Module*             pModule,
+    BOOL                isBaseTypeSig,
+    mdToken*            pTypeDefToken,                  /* = NULL */
+    Module**            ppTypeDefModule,                /* = NULL */
+    mdToken*            pParentTypeDefOrRefOrSpecToken) /* = NULL */
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    CorElementType elementType = ELEMENT_TYPE_MAX;
+
+    IfFailThrow(CorSigUncompressElementType_EndPtr(pSig, pEndSig, &elementType));
+
+    // System.Object is always eligible as a base type since all reference types derive from it
+    if (isBaseTypeSig && elementType == ELEMENT_TYPE_OBJECT)
+    {
+        return TRUE;
+    }
+
+    if (elementType == ELEMENT_TYPE_GENERICINST)
+    {
+        IfFailThrow(CorSigUncompressElementType_EndPtr(pSig, pEndSig, &elementType));
+    }
+
+    // Only classes are eligible (no inheritance with value types)
+    if (elementType != ELEMENT_TYPE_CLASS)
+    {
+        return FALSE;
+    }
+
+    mdToken typeRefOrDefToken;
+    IfFailThrow(CorSigUncompressToken_EndPtr(pSig, pEndSig, &typeRefOrDefToken));
+
+    // Ensure we are working with the typedef token
+    mdToken typeDefToken;
+    Module* pFoundModule;
+    if (!ClassLoader::ResolveTokenToTypeDefThrowing(pModule, typeRefOrDefToken, &pFoundModule, &typeDefToken))
+    {
+        return FALSE;
+    }
+
+    mdToken parentTypeDefOrRefOrSpecToken = mdTokenNil;
+    if (!isBaseTypeSig)
+    {
+        DWORD attr;
+        IfFailThrow(pFoundModule->GetMDImport()->GetTypeDefProps(typeDefToken, &attr, &parentTypeDefOrRefOrSpecToken));
+
+        // Check the validity of the parent type token if this is a signature of a derived type.
+        if (!pFoundModule->GetMDImport()->IsValidToken(parentTypeDefOrRefOrSpecToken))
+        {
+            return FALSE;
+        }
+    }
+
+    if (ppTypeDefModule != NULL)
+        *ppTypeDefModule = pFoundModule;
+    if (pTypeDefToken != NULL)
+        *pTypeDefToken = typeDefToken;
+    if (pParentTypeDefOrRefOrSpecToken != NULL)
+        *pParentTypeDefOrRefOrSpecToken = parentTypeDefOrRefOrSpecToken;
+
+    return TRUE;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// Compute the base type token (it will either be a typedef or typeref token), and return the
+// module where that base type token is declared.
+//
+// static
+BOOL MetaSig::GetBaseTypeTokenAndModule(
+    mdToken     tk,
+    Module*     pModule,
+    mdToken*    pBaseTypeDefOrRefToken,
+    Module**    ppBaseTypeTokenModule)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        PRECONDITION(CheckPointer(pModule));
+        PRECONDITION(CheckPointer(pBaseTypeDefOrRefToken));
+        PRECONDITION(CheckPointer(ppBaseTypeTokenModule));
+        PRECONDITION(TypeFromToken(tk) == mdtTypeRef || TypeFromToken(tk) == mdtTypeDef);
+        MODE_ANY;
```

Comment:
It feels dangerous to run all this with `MODE_ANY`. Can it be restricted to `MORE_PREEMPTIVE` ?

Reply:
The main comparison algorithm is implemented as MODE_ANY (`CompareElementType()`). Would there be any side effect from making the new APIs MODE_PREEMPTIVE, while the caller `CompareElementType()` is MODE_ANY?

Reply:
Does that main comparison algorithm need to be `MODE_ANY` ? Can we flip the whole thing to MODE_PREEMPTIVE ?

Reply:
I can't say with certainty. Maybe @davidwrighton can help here